### ISSUE TITLE
Output of GenSLMs

### ISSUE THREAD

**Issue by Cassie818** [ramanathanlab/genslm#106]
<img width="754" alt="Screenshot 2024-06-12 at 2 53 42 PM" src="https://github.com/ramanathanlab/genslm/assets/80612275/dcdd782a-abb4-4d13-8080-b1c99d37e1b0">

I’m using the scripts provided above to extract the output embedding of GenSLMs. The paper said “GenSLMs take nucleotide sequences of SARS-CoV-2 genomes as input and build LLMs that learn a semantic embedding of individual codons, which can then be translated to the 29 individual protein sequences that are encoded by the virus”. Hence, the outputs are the probabilities of amino acids, aren't they? However, the output shape doesn't match the input nucleotide sequences. If we want to study the specific output possibilities for each amino acid in a protein sequence, then we can't take the average value. 
<img width="452" alt="image" src="https://github.com/ramanathanlab/genslm/assets/80612275/f30a5880-1c63-48e1-886f-c0d02017a945"> 
What can I do to extract the embedding output shown in the paper?  Or can you explain in more detail what this output actually is?Thanks for your help!

**Comment by KPHippe** [comment]
Hi, you are correct in that the hidden states do not have the dimensionality to reconstruct the probabilities of the nucleotides. To reconstruct the probabilities of the nucleotides you will have to look at the `outputs.logits` as this is the output of the language modeling head of the model. You can find more information on the outputs class [here](https://huggingface.co/docs/transformers/en/main_classes/output#transformers.modeling_outputs.CausalLMOutputWithPast) The embeddings (coming from the hidden states of the model) are merely an information dense representation of the sequence, but cannot be directly used to decode the codon. 

When encoding the nucleotide sequence of the SARS-CoV-2 genome, we input the nucleotide sequence and they are split into 3-mers. This is not an amino acid representation ($4^3=64$ possible tokens), and is only the codon representation if the input is fed in such that all proteins are in phase by the 3-mer tokenization.
